qiskit/providers/aer/openpulse/hamiltonian_model.py

```diff
@@ -0,0 +1,344 @@
+            H_full += hpart[0].full() * eval(hpart[1])
+
+        evals, estates = la.eigh(H_full)
+
+        eval_mapping = []
+        for ii in range(len(evals)):
+            eval_mapping.append(np.argmax(np.abs(estates[:, ii])))
+
+        # Remap eigenvalues and eigenstates
+        evals_mapped = evals.copy()
+        estates_mapped = estates.copy()
+
+        for ii, val in enumerate(eval_mapping):
+            evals_mapped[val] = evals[ii]
+            estates_mapped[:, val] = estates[:, ii]
+
+        self._evals = evals_mapped
+        self._estates = estates_mapped
+        self._h_diag = np.ascontiguousarray(np.diag(H_full).real)
+
+
+def _first_excited_state(qubit_idx, dim_qub):
+    """
+    Returns the vector corresponding to all qubits in the 0 state, except for
+    qubit_idx in the 1 state.
+
+    Assumption: the keys in dim_qub consist exactly of the str version of the int
+                in range(len(dim_qub)). They don't need to be in order, but they
+                need to be of this format
+
+    Parameters:
+        qubit_idx (int): the qubit to be in the 1 state
+
+        dim_qub (dict): a dictionary with keys being qubit index as a string, and
+                        value being the dimension of the qubit
+
+    Returns:
+        vector: the state with qubit_idx in state 1, and the rest in state 0
+    """
+    vector = np.array([1.])
+
+    # iterate through qubits, tensoring on the state
+    for idx, dim in enumerate(dim_qub):
+        new_vec = np.zeros(dim)
+        if idx == qubit_idx:
+            new_vec[1] = 1
+        else:
+            new_vec[0] = 1
+
+        vector = np.kron(new_vec, vector)
+
+    return vector
+
+
+def _eval_for_max_espace_overlap(u, evals, evecs, decimals=14):
+    """ Given an eigenvalue decomposition evals, evecs, as output from
+    get_diag_hamiltonian, returns the eigenvalue from evals corresponding
+    to the eigenspace that the vector vec has the maximum overlap with.
+
+    Parameters:
+        u (numpy.array): the vector of interest
+
+        evals (numpy.array): list of eigenvalues
+
+        evecs (numpy.array): eigenvectors corresponding to evals
+
+        decimals (int): rounding option, to try to handle numerical
+                        error if two evals should be the same but are
+                        slightly different
+
+    Returns:
+        eval: eigenvalue corresponding to eigenspace for which vec has
+              maximal overlap
+
+    Raises:
+    """
+
+    # get unique evals (with rounding for numerical error)
+    rounded_evals = evals.copy().round(decimals=decimals)
+    unique_evals = np.unique(rounded_evals)
+
+    # compute overlaps to unique evals
+    overlaps = np.zeros(len(unique_evals))
+    for idx, val in enumerate(unique_evals):
+        overlaps[idx] = _proj_norm(evecs[:, val == rounded_evals], u)
+
+    # return eval with largest overlap
+    return unique_evals[np.argmax(overlaps)]
+
+
+def _proj_norm(A, b):
```

Comment:
This is a war I lose long time about.. but here I go: I guess that `A` and `b` are very common in math contexts, but extremely ugly for readability, would some other naming be acceptable here? :)

Reply:
I do find it readable purely due to the commonality to math contexts, as well as it primarily being a call to `la.lstsq`, which uses `A` and `b` as arguments. I can't think of anything else, but I may just be too heavily indoctrinated :P .

Reply:
Change to `mat` `vec` to make Juan happy